File: java/low-level-design/src/com/educosys/parkinglot/ParkingLot.java
package com.educosys.parkinglot;

import java.util.List;

public class ParkingLot {
    private static ParkingLot instance;
    private List<Level> levels;

    public static ParkingLot getInstance()
    {
        if(instance==null)
        {
            instance=new ParkingLot();
        }
        return instance;
    }

    public void addLevels(Level level)
    {
        this.levels.add(level);
    }

    public boolean parkVehicle(Vehicle vehicle)
    {
        for(Level level: levels)
        {
            if(level.parkVehicle(vehicle))
                    return true;
        }

        return false;
    }

    public boolean unparkVehicle(Vehicle vehicle)
    {
        for(Level level: levels)
        {
            if(level.unparkVehicle(vehicle))
                return true;
        }

        return false;
    }

    public void displayParking()
    {
        for(Level level: levels)
        {
            level.displayParking();
        }
    }

}

File: c++/parking-lot-lld/ParkingLot.cpp
#include<bits/stdc++.h>
#include "Level.h"
#include "Vehicle.h"
#include "ParkingLot.h"

using namespace std;

ParkingLot* ParkingLot::instance = NULL;

ParkingLot::ParkingLot(){}

ParkingLot* ParkingLot::getInstance()
{
    if(instance==NULL){
        instance = new ParkingLot();
    }
    return instance;
}

void ParkingLot::addLevels(Level *level){
    this->levels.push_back(level);
}

bool ParkingLot::parkVehicle(Vehicle *vehicle)
{
    for(auto floor: levels)
    {
        if(floor->parkVehicle(vehicle) == true)
            return true;
    }

    return false;
}

bool ParkingLot::unparkVehicle(Vehicle *vehicle)
{
    for(auto floor: levels)
    {
        if(floor->unparkVehicle(vehicle) == true)
            return true;
    }
    return false;
}

void ParkingLot::displayParkingLot()
{
    for(auto it: levels)
    {
        it->displayLevels();
    }
}
